Source/Core/InputCommon/ControlReference/ExpressionParser.cpp

```diff
@@ -374,6 +471,90 @@ class VariableExpression : public Expression
   ControlState* m_value_ptr{};
 };
 
+class HotkeyExpression : public Expression
+{
+public:
+  HotkeyExpression(std::vector<std::unique_ptr<ControlExpression>> inputs)
+      : m_modifiers(std::move(inputs))
+  {
+    m_final_input = std::move(m_modifiers.back());
+    m_modifiers.pop_back();
+  }
+
+  ControlState GetValue() const override
+  {
+    const bool modifiers_pressed = std::all_of(m_modifiers.begin(), m_modifiers.end(),
+                                               [](const std::unique_ptr<ControlExpression>& input) {
+                                                 return input->GetValue() > CONDITION_THRESHOLD;
+                                               });
+
+    const auto final_input_state = m_final_input->GetValueIgnoringSuppression();
+
+    if (modifiers_pressed)
+    {
+      if (final_input_state < CONDITION_THRESHOLD)
+      {
+        if (!m_suppressor)
+          EnableSuppression();
+
+        m_is_ready = true;
+      }
+
+      // Ignore suppression of our own modifiers. This also allows superset modifiers to function.
+      const bool is_suppressed = s_hotkey_suppressions.IsSuppressedIgnoringModifiers(
+          m_final_input->GetInput(), m_modifiers);
+
+      // If some other hotkey suppressed us, require a release of final input to be ready again.
+      if (is_suppressed)
+        m_is_ready = false;
+
+      // Our modifiers are active. Pass through the final input.
+      if (m_is_ready)
+        return final_input_state;
+    }
+    else
```

Comment:
Not sure I fully understand this yet, but shouldn't the HotkeyExpression be blocked if the "final_input" (the only non modifier input) is pressed before the modifiers? And stay blocked until it is reseased.
So that if you press O an then CTRL while O is still pressed, it wouldn't work. This doesn't seem to be the way now.

There is no need for time stamps on inputs, just make sure the modifiers were pressed in the same frame or any frame before (by frame, I mean a refresh of the cached input values from the control devices).

Reply:
Yes. If the "final_input" is pressed before the modifiers then `m_is_ready` will not be set.

Reply:
This is weird though, the different modifiers aren't required to be pressed within any order or time, while the final_input needs to be pressed at least one frame after the rest. I guess this is mostly for the Hotkeys and you'd be insane to use a modifier as a game control, but if Dolphin was running slow then it might be hard to get an input with an hotkey to pass. I think they should also be accepted if both the modifiers and the final_input were pressed in the same frame.

Another problem with this is this:
-Press O
-Press CTRL
-Release O
-Press O again
In this case, you'd expect the final CTRL+O to be accepted, but it wouldn't. That's why i had not looked inside the "if (modifiers_pressed)" condition, I expected the "final_input" blocking to be handled outside of it.

Reply:
The final "CTRL+O" **will** be accepted in your example situation. `m_is_ready` will be set when "O" is released.

Reply:
I would agree that pressing modifiers and the final input in the same frame could/should be accepted.
I think this might require an additional state variable.

Reply:
More like wronglythinking

Reply:
No, if you just move the code that sets `m_is_ready` outside of `modifiers_pressed`, and rename it `m_is_blocked`, it should work.
Basically you block everything if `final_input` is pressed before `modifiers_pressed`, until it's released.

Reply:
I'll make the change when I get around to giving this a rebase.